File: bad-tournament/app/types/tournament_category.ts
export type GenderType = "men" | "women" | "mixed";
export type EventType = "singles" | "doubles" | "mixed_doubles";
export type FormatType = "elimination" | "league" | "league_to_tournament";

export type TournamentCategory = {
  id: number;
  tournament_id: number;
  gender_type: GenderType;
  event_type: EventType;
  age_type: string;
  rank: string;
  format_type: FormatType;
  max_participants: number | null;
  group_size: number | null;
  group_count: number | null;
  advance_count: number | null;
  has_third_place: boolean;
  created_at: string;
  updated_at: string;
};

export type CreateTournamentCategoryParams = {
  gender_type: GenderType;
  event_type: EventType;
  age_type: string;
  rank: string;
  format_type: FormatType;
  max_participants?: number | null;
  group_size?: number | null;
  group_count?: number | null;
  advance_count?: number | null;
  has_third_place?: boolean;
};

// 表示用ラベル
export const GENDER_TYPE_LABELS: Record<GenderType, string> = {
  men: "男子",
  women: "女子",
  mixed: "混合",
};

export const EVENT_TYPE_LABELS: Record<EventType, string> = {
  singles: "シングルス",
  doubles: "ダブルス",
  mixed_doubles: "ミックスダブルス",
};

export const FORMAT_TYPE_LABELS: Record<FormatType, string> = {
  elimination: "トーナメント",
  league: "リーグ",
  league_to_tournament: "予選リーグ → 本戦トーナメント",
};

// 一括作成用の下書きカテゴリー型（フォーム入力値を string で保持）
export type DraftCategoryForm = {
  localId: string; // React key 用のローカルID
  gender_type: GenderType;
  event_type: EventType;
  age_type: string;
  rank: string;
  format_type: FormatType | "";
  max_participants: string;
  group_size: string;
  group_count: string;
  advance_count: string;
  has_third_place: boolean;
};

File: backend/app/models/tournament_category.rb
class TournamentCategory < ApplicationRecord
  belongs_to :tournament
  has_many :teams,          dependent: :destroy
  has_many :league_groups,  dependent: :destroy
  has_many :matches,        dependent: :destroy

  enum :gender_type, { men: 0, women: 1, mixed: 2 }

  enum :event_type, { singles: 0, doubles: 1, mixed_doubles: 2 }

  enum :format_type, { elimination: 0, league: 1, league_to_tournament: 2 }

  validates :gender_type,  presence: true
  validates :event_type,   presence: true
  validates :format_type,  presence: true
  validates :rank,         presence: true
  # age_type は任意（年齢区分がない大会も多いため）

  with_options if: :league_needed? do
    validates :group_size, presence: true,
                           numericality: { only_integer: true, greater_than_or_equal_to: 1 }
  end

  with_options if: :finals_needed? do
    validates :advance_count, presence: true,
                              numericality: { only_integer: true, greater_than_or_equal_to: 1 }
  end

  with_options if: :tournament_needed? do
    validates :max_participants, presence: true,
                                 numericality: { only_integer: true, greater_than_or_equal_to: 2 }
  end

  private

  def league_needed?
    league? || league_to_tournament?
  end

  def finals_needed?
    league_to_tournament?
  end

  def tournament_needed?
    elimination? || league_to_tournament?
  end
end
